### PR TITLE
[#4386] Fix(spark-connector): spark connector build failed with scala2.13

### PR DESCRIPTION
**Pull request by caican00**
### What changes were proposed in this pull request?
Fixed spark connector build failed with scala2.13.

### Why are the changes needed?

Fix:  https://github.com/apache/gravitino/issues/4386

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
1. build with scala2.12 and scala2.13 locally.
2. add a new job about building spark-connector in github pipeline to test scala2.13

### CODE REVIEW

**Comment by caican00** [comment]
cc @FANNG1

**Comment by FANNG1** [comment]
Could you run pass the IT?

**Comment by caican00** [comment]
> ### What changes were proposed in this pull request?
> Fixed spark connector build failed with scala2.13.
> 
> ### Why are the changes needed?
> Fix: #4386
> 
> ### Does this PR introduce _any_ user-facing change?
> No.
> 
> ### How was this patch tested?
> build with scala2.12 and scala2.13 locally.
> 
> `./gradlew clean :spark-connector:spark-runtime-3.3:build -PscalaVersion=2.12 -PjdkVersion=17 -x test`
> 
> ```
> 
> ./gradlew clean :spark-connector:spark-runtime-3.4:build -PscalaVersion=2.12 -PjdkVersion=17 -x test
> 
> ./gradlew clean :spark-connector:spark-runtime-3.4:build -PscalaVersion=2.13 -PjdkVersion=17 -x test
> ```
> 
> ```
> ./gradlew clean :spark-connector:spark-runtime-3.5:build -PscalaVersion=2.12 -PjdkVersion=17 -x test
> 
> ./gradlew clean :spark-connector:spark-runtime-3.5:build -PscalaVersion=2.13 -PjdkVersion=17 -x test
> ```

got it.

**Comment by jerryshao** [comment]
@FANNG1 can you please help to review this.

**Comment by caican00** [comment]
> Could you run pass the IT? seems there are still some problems in runtime.

got it.

**Comment by caican00** [comment]
> Could you run pass the IT? seems there are still some problems in runtime.

image

i tested it with scala2.13 and java17 locally, and run successfully. 
could you provide the way to reproduce? @FANNG1

**Comment by caican00** [comment]
@FANNG1 should we add the scala2.13 jobs to the pipeline?

File: .github/workflows/spark-integration-test.yml [hunk 1]
```diff
@@ -58,7 +58,7 @@ jobs:
       matrix:
         architecture: [linux/amd64]
         java-version: [ 8, 11, 17 ]
-        scala-version: [ 2.12 ]
+        scala-version: [ 2.12, 2.13 ]
```
Comment (caican00) [inline .github/workflows/spark-integration-test.yml:61]: @FANNG1 should we add the scala2.13 jobs to the pipeline?

**Comment by FANNG1** [comment]
> > Could you run pass the IT? seems there are still some problems in runtime.
> 
> image
> 
> i tested it with scala2.13 and java17 locally, and ran successfully. could you provide the way to reproduce? @FANNG1

I could run successfully now, and couldn't reproduce it.

File: .github/workflows/spark-integration-test.yml (hunk 1 above)
Comment (FANNG1) [reply .github/workflows/spark-integration-test.yml:61]: There are too many pipelines for now, let's check the test result of the ci

Comment (FANNG1) [reply .github/workflows/spark-integration-test.yml:61]: There are test errors in the ci, could you fix it?

Comment (caican00) [reply .github/workflows/spark-integration-test.yml:61]: > There are test errors in the ci, could you fix it?

sure

File: .github/workflows/spark-integration-test.yml [hunk 2]
```diff
@@ -92,9 +92,13 @@ jobs:
         run: |
           if [ "${{ matrix.scala-version }}" == "2.12" ];then 
             ./gradlew -PskipTests -PtestMode=${{ matrix.test-mode }} -PjdkVersion=${{ matrix.java-version }} -PscalaVersion=${{ matrix.scala-version }} -PskipDockerTests=false :spark-connector:spark-3.3:test --tests "org.apache.gravitino.spark.connector.integration.test.**"
+            ./gradlew -PskipTests -PtestMode=${{ matrix.test-mode }} -PjdkVersion=${{ matrix.java-version }} -PscalaVersion=${{ matrix.scala-version }} -PskipDockerTests=false :spark-connector:spark-3.4:test --tests "org.apache.gravitino.spark.connector.integration.test.**"
+            ./gradlew -PskipTests -PtestMode=${{ matrix.test-mode }} -PjdkVersion=${{ matrix.java-version }} -PscalaVersion=${{ matrix.scala-version }} -PskipDockerTests=false :spark-connector:spark-3.5:test --tests "org.apache.gravitino.spark.connector.integration.test.**"
+          fi
+          if [ "${{ matrix.scala-version }}" == "2.13" ];then 
+            ./gradlew -PtestMode=${{ matrix.test-mode }} -PjdkVersion=${{ matrix.java-version }} -PscalaVersion=${{ matrix.scala-version }} :spark-connector:spark-3.4:test
```
Comment (caican00) [inline .github/workflows/spark-integration-test.yml:99]: I think we must provide a way to test for compatibility with scala2.13, and if adding scala2.13 will make the pipelines too slow, we can skip docker tests. WDYT? @FANNG1

File: spark-connector/spark-common/src/main/java/org/apache/gravitino/spark/connector/utils/ConnectorUtil.java [hunk 3]
```diff
@@ -37,4 +41,11 @@ public static String removeDuplicateSparkExtensions(
         .reduce((element1, element2) -> element1 + ConnectorConstants.COMMA + element2)
         .orElse("");
   }
+
+  @VisibleForTesting
+  public static List<String> toJavaList(Seq<String> seq) {
```
Comment (FANNG1) [inline spark-connector/spark-common/src/main/java/org/apache/gravitino/spark/connector/utils/ConnectorUtil.java:46]: is it possible seq is null?

Comment (caican00) [reply spark-connector/spark-common/src/main/java/org/apache/gravitino/spark/connector/utils/ConnectorUtil.java:46]: ooh, i will add a check here.

File: .github/workflows/spark-integration-test.yml (hunk 2 above)
Comment (FANNG1) [reply .github/workflows/spark-integration-test.yml:99]: How about add spark connector build and UT with scala2.13  in `.github/workflows/build.yml`?

File: spark-connector/spark-common/src/main/java/org/apache/gravitino/spark/connector/iceberg/extensions/IcebergExtendedDataSourceV2Strategy.java [hunk 4]
```diff
@@ -81,7 +84,7 @@ public Seq<SparkPlan> apply(LogicalPlan plan) {
     } else if (plan instanceof CreateOrReplaceBranch) {
       CreateOrReplaceBranch createOrReplaceBranch = (CreateOrReplaceBranch) plan;
       return IcebergCatalogAndIdentifier.buildCatalogAndIdentifier(
-              spark, createOrReplaceBranch.table())
+              spark, createOrReplaceBranch.table().toIndexedSeq())
```
Comment (FANNG1) [inline spark-connector/spark-common/src/main/java/org/apache/gravitino/spark/connector/iceberg/extensions/IcebergExtendedDataSourceV2Strategy.java:87]: why changing `toIndexedSeq`?

File: spark-connector/spark-common/src/main/java/org/apache/gravitino/spark/connector/utils/ConnectorUtil.java [hunk 5]
```diff
@@ -37,4 +41,11 @@ public static String removeDuplicateSparkExtensions(
         .reduce((element1, element2) -> element1 + ConnectorConstants.COMMA + element2)
         .orElse("");
   }
+
+  @VisibleForTesting
```
Comment (FANNG1) [inline spark-connector/spark-common/src/main/java/org/apache/gravitino/spark/connector/utils/ConnectorUtil.java:45]: @VisibleForTesting ?

File: .github/workflows/spark-integration-test.yml (hunk 2 above)
Comment (caican00) [reply .github/workflows/spark-integration-test.yml:99]: ok for me.

File: spark-connector/spark-common/src/main/java/org/apache/gravitino/spark/connector/utils/ConnectorUtil.java (hunk 5 above)
Comment (caican00) [reply spark-connector/spark-common/src/main/java/org/apache/gravitino/spark/connector/utils/ConnectorUtil.java:45]: any problem here?

Comment (FANNG1) [reply spark-connector/spark-common/src/main/java/org/apache/gravitino/spark/connector/utils/ConnectorUtil.java:45]: it's used in different packages, should remove `VisibleForTesting`

File: spark-connector/spark-common/src/main/java/org/apache/gravitino/spark/connector/iceberg/extensions/IcebergExtendedDataSourceV2Strategy.java (hunk 4 above)
Comment (caican00) [reply spark-connector/spark-common/src/main/java/org/apache/gravitino/spark/connector/iceberg/extensions/IcebergExtendedDataSourceV2Strategy.java:87]:
```
error: apply(LogicalPlan) in IcebergExtendedDataSourceV2Strategy cannot override apply(LogicalPlan) in GenericStrategy
  public Seq<SparkPlan> apply(LogicalPlan plan) {
                        ^
  return type scala.collection.Seq<SparkPlan> is not compatible with scala.collection.immutable.Seq<SparkPlan>

```

The `type scala.Seq` in scala2.13 represents the  `scala.collection.immutable.Seq`, `toIndexedSeq` can change the return type to `scala.collection.immutable.Seq`.
and `IndexedSeq` is a sub `trait` of `scala.collection.immutable.Seq`

File: spark-connector/spark-common/src/main/java/org/apache/gravitino/spark/connector/utils/ConnectorUtil.java (hunk 3 above)
Comment (caican00) [reply spark-connector/spark-common/src/main/java/org/apache/gravitino/spark/connector/utils/ConnectorUtil.java:44]: fixed.

File: .github/workflows/spark-integration-test.yml (hunk 2 above)
Comment (caican00) [reply .github/workflows/spark-integration-test.yml]: updated.

File: spark-connector/spark-common/src/main/java/org/apache/gravitino/spark/connector/utils/ConnectorUtil.java (hunk 5 above)
Comment (caican00) [reply spark-connector/spark-common/src/main/java/org/apache/gravitino/spark/connector/utils/ConnectorUtil.java]: removed.

**Comment by caican00** [comment]
@FANNG1 could you help review again?

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -107,3 +107,41 @@ jobs:
...
+        with:
+          java-version: 8
+          distribution: 'temurin'
+          cache: 'gradle'
+
+      - name: Free up disk space
+        run: |
+          dev/ci/util_free_space.sh
+
+      - name: Build with Gradle
```
Comment (FANNG1) [inline .github/workflows/build.yml:132]: How about adding a separate step in `build` named like `Build spark connector with scala 2.13` to build and test scala 2.13 packages.

**Comment by FANNG1** [comment]
> @FANNG1 could you help review again?

just one comment

File: .github/workflows/build.yml (hunk 6 above)
Comment (caican00) [reply .github/workflows/build.yml:132]: @FANNG1 just test scala2.13, not considering scala2.12?

Comment (caican00) [reply .github/workflows/build.yml]: updated, only support scala2.13

**Comment by caican00** [comment]
> > @FANNG1 could you help review again?
> 
> just one comment

updated cc @FANNG1

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -98,6 +98,11 @@ jobs:
       - name: Build with Gradle
         run: ./gradlew build -PskipITs -PjdkVersion=${{ matrix.java-version }} -PskipDockerTests=false -x :clients:client-python:build
 
+      - name: Build spark connector with scala 2.13
+        run: |
+          ./gradlew -PscalaVersion=2.13 :spark-connector:spark-3.4:test
```
Comment (FANNG1) [inline .github/workflows/build.yml:103]: Could you use the command below to skip ITs?
```
 ./gradlew  :spark-connector:spark-3.4:test -PskipITs -PjdkVersion=${{ matrix.java-version }} -PskipDockerTests=false -PscalaVersion=2.13 
 ./gradlew  :spark-connector:spark-3.5:test -PskipITs -PjdkVersion=${{ matrix.java-version }} -PskipDockerTests=false -PscalaVersion=2.13 

```

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -98,6 +98,11 @@ jobs:
       - name: Build with Gradle
         run: ./gradlew build -PskipITs -PjdkVersion=${{ matrix.java-version }} -PskipDockerTests=false -x :clients:client-python:build
 
+      - name: Build spark connector with scala 2.13
```
Comment (FANNG1) [inline .github/workflows/build.yml:101]: There may be one problem in current pipeline, if `Build with Gradle` failed but `Build spark` successful, no log is updated.

Comment (FANNG1) [inline .github/workflows/build.yml:101]: @jerryshao @xunliu @yuqi1129  please help to review the pipeline changes.

File: .github/workflows/build.yml (hunk 7 above)
Comment (caican00) [reply .github/workflows/build.yml:103]: the `skipDockerTests` param is true by default.

File: .github/workflows/build.yml (hunk 8 above)
Comment (caican00) [reply .github/workflows/build.yml:101]: why not add it in a new stage

**Comment by caican00** [comment]
```
CatalogDorisDriverIT > initializationError FAILED
    org.testcontainers.containers.ContainerLaunchException: Container startup failed for image datastrato/gravitino-ci-doris:0.1.5
        at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:361)
        at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:334)
        at org.apache.gravitino.integration.test.container.BaseContainer.start(BaseContainer.java:154)
        at org.apache.gravitino.integration.test.container.DorisContainer.start(DorisContainer.java:77)
        at org.apache.gravitino.integration.test.container.ContainerSuite.startDorisContainer(ContainerSuite.java:236)
        at org.apache.gravitino.catalog.doris.integration.test.CatalogDorisIT.startup(CatalogDorisIT.java:119)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)

```

please help re-trigger the pipeline, Thanks @FANNG1

File: .github/workflows/build.yml (hunk 8 above)
Comment (FANNG1) [reply .github/workflows/build.yml:101]: I couldn't understand `new stage`

Comment (caican00) [reply .github/workflows/build.yml:101]: @FANNG1 what I mean is to add it as a new job for the build workflow, like this:

```

  spark-connector-build:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        scala-version: [ 2.12, 2.13 ]
    timeout-minutes: 30
    needs: changes
    if: needs.changes.outputs.source_changes == 'true'
    steps:
      - uses: actions/checkout@v3

      - uses: actions/setup-java@v4
        with:
          java-version: 8
          distribution: 'temurin'
          cache: 'gradle'

      - name: Free up disk space
        run: |
          dev/ci/util_free_space.sh

      - name: Build with Gradle
        run: |
          if [ "${{ matrix.scala-version }}" == "2.12" ];then 
            ./gradlew -PscalaVersion=${{ matrix.scala-version }} :spark-connector:spark-3.3:test
          fi
          ./gradlew -PscalaVersion=${{ matrix.scala-version }} :spark-connector:spark-3.4:test
          ./gradlew -PscalaVersion=${{ matrix.scala-version }} :spark-connector:spark-3.5:test

      - name: Upload unit tests report
        uses: actions/upload-artifact@v3
        if: failure()
        with:
          name: unit test report
          path: |
            build/reports
            spark-connector/**/*.log

```

after the `changes` succeeded, the `spark-connector-build` can be executed.

Comment (FANNG1) [reply .github/workflows/build.yml]: What my concern is this will add more extra steps  to make build time much longer.

Comment (caican00) [reply .github/workflows/build.yml]: imo, it is insignificant compared to the time taking for all the jobs

Comment (FANNG1) [reply .github/workflows/build.yml]: so what's the benefit of using a job not a step?

Comment (caican00) [reply .github/workflows/build.yml]: > so what's the benefit of using a job not a step?

@FANNG1 It does not affect the job named `build` and they can be built in parallel.

**Comment by jerryshao** [comment]
Is everything done, can we move fast to make this merged today?

File: .github/workflows/build.yml (hunk 8 above)
Comment (FANNG1) [reply .github/workflows/build.yml]: ok, let's use a job test spark connector with scala 2.13, please note that seems no need to test scala 2.12 again

**Comment by caican00** [comment]
> Is everything done, can we move fast to make this merged today? @FANNG1 @caican00 , thanks.

https://github.com/apache/gravitino/pull/4403#discussion_r1713127894
@FANNG1 Do you have any further comments? after https://github.com/apache/gravitino/pull/4468 is merged, we can re-trigger the pipelines again.

File: .github/workflows/build.yml (hunk 8 above)
Comment (caican00) [reply .github/workflows/build.yml]: > please note that seems no need to test scala 2.12 again
 yeah, there is no scala2.12.

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -68,6 +68,39 @@ jobs:
...
+          distribution: 'temurin'
+          cache: 'gradle'
+
+      - name: Free up disk space
+        run: |
+          dev/ci/util_free_space.sh
+
+      - name: Build with Scala2.13
+        run: |
+          ./gradlew -PscalaVersion=2.13 -PskipITs -PskipDockerTests=false :spark-connector:spark-3.4:test
```
Comment (FANNG1) [inline .github/workflows/build.yml:92]: please use 
```
          ./gradlew  :spark-connector:spark-3.4:build -PscalaVersion=2.13 -PskipITs -PskipDockerTests=false
          ./gradlew  :spark-connector:spark-3.5:build -PscalaVersion=2.13 -PskipITs -PskipDockerTests=false 
```

File: .github/workflows/build.yml (hunk 8 above)
Comment (caican00) [reply .github/workflows/build.yml]: > > so what's the benefit of using a job not a step?
> 
> @FANNG1 It does not affect the job named `build` and they can be built in parallel.

image
cc @FANNG1

File: .github/workflows/build.yml (hunk 9 above)
Comment (caican00) [reply .github/workflows/build.yml]: updated.

File: .github/workflows/build.yml (hunk 8 above)
Comment (FANNG1) [reply .github/workflows/build.yml]: cool

**Comment by FANNG1** [comment]
LGTM

**Review by FANNG1 — APPROVED** [review]

**Comment by FANNG1** [comment]
@caican00 merged to main, thanks for your work

**Pull request merged**